pulsar-client/src/main/java/org/apache/pulsar/client/impl/PersistentAcknowledgmentsGroupingTracker.java

```diff
@@ -330,16 +462,41 @@ public void flush() {
             return;
         }
 
+        if (ackResponseEnabled) {
+            this.lock.writeLock().lock();
+            try {
+                flushAsync(cnx);
+            } finally {
+                this.lock.writeLock().unlock();
+            }
+        } else {
+            flushAsync(cnx);
+        }
+    }
+
+    private void flushAsync(ClientCnx cnx) {
         boolean shouldFlush = false;
         if (cumulativeAckFlushRequired) {
-            newAckCommand(consumer.consumerId, lastCumulativeAck, lastCumulativeAckSet, AckType.Cumulative, null, Collections.emptyMap(), cnx, false /* flush */, -1, -1);
+            if (ackResponseEnabled) {
+                long requestId = consumer.getClient().newRequestId();
+                ByteBuf cmd = Commands.newAck(consumer.consumerId, lastCumulativeAck.messageId.ledgerId,
+                        lastCumulativeAck.messageId.getEntryId(), lastCumulativeAck.bitSetRecyclable,
+                        AckType.Cumulative, null, Collections.emptyMap(), requestId);
+                cnx.newAckForResponseWithFuture(cmd, requestId, currentCumulativeAckFuture);
+                this.currentCumulativeAckFuture = new TimedCompletableFuture<>();
+            } else {
+                ByteBuf cmd = Commands.newAck(consumer.consumerId, lastCumulativeAck.messageId.ledgerId,
+                        lastCumulativeAck.messageId.getEntryId(), lastCumulativeAck.bitSetRecyclable,
+                        AckType.Cumulative, null, Collections.emptyMap(), -1);
+                cnx.ctx().write(cmd, cnx.ctx().voidPromise());
```

Comment:
Please optimize duplicate code

Reply:
if we optimize the code, we will judge the ackResponseEnabled twice, because get requestId then new cmd and then write. get requestId and write are different from normal ack and ack response. the middle operation newAckComand can't be optimized.